### ISSUE TITLE
sharedfp/sm component broken on master with threading changes

### ISSUE THREAD

**Issue by edgargabriel** [open-mpi/ompi#8464]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

master

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

source code from git clone

### Please describe the system on which you are running

linux OpenSuSE 15.2, but most likely not limited to it.

* Operating system/version: 
* Computer hardware: 
* Network type: 

-----------------------------

## Details of the problem

Due to some recent changes on master, the ompi/sharedfp/sm component does not pass the configure test anymore and does not compile anymore either. @ggouaillardet identified the configure issue was due to a missing dependence to pthread. 

I fixed the configure script of sharedfp/sm to accommodate for that, and we pass now the configure test again. The code for this fix can be found in the  edgargabriel/ompi  repository, in the topic/sharedfp-sm-configure-fix branch.

However, the code suddenly doesn't compile anymore as well, with an obscure error that I do not understand where it is coming from (code of the component is unchanged compared to the 4.1 and 4.0 releases). I need help from somebody familiar with the recent threading changes. The component is probably a bit unusual for ompi in that it calls mmap and semaphore operations directly (i.e. not through opal operations). If we need to change this, I am more than happy to help. but I need some guidance on it.

**Comment by hppritcha** [comment]
i'll discuss this with argobot crew and see how to proceed.

**Comment by shintaro-iwasaki** [comment]
I'm an Argobots crew.  If my understanding is wrong, please be patient and correct me.

First, I could not configure Open MPI to automatically compile this `sharedfp/sm`.  I cannot find how to enable this even before the MCA/OPAL/threads etc ( https://github.com/open-mpi/ompi/commit/b3f0bc549072cf28b61076efad5cbe045d217941).  To compile it, I went this directory and explicitly `make` it.
```sh
cd ompi/mca/sharedfp/sm
make
```

Then the current Open MPI main branch shows the following:
```sh
.../ompi/mca/sharedfp/sm$ make
...
  CC       sharedfp_sm_file_open.lo
sharedfp_sm_file_open.c:210:6: error: expected identifier or ‘(’ before ‘else’
  210 |     }else{
      |      ^~~~
sharedfp_sm_file_open.c:218:5: warning: data definition has no type or storage class
  218 |     err = comm->c_coll->coll_barrier (comm, comm->c_coll->coll_barrier_module );
      |     ^~~
...
```
Assuming that this is an error you mentioned.

1. At some points (maybe when MCA/OPAL/threads is introduced but maybe later), Open MPI does not add `-lpthread` for `LIBS` by default, so the configuration script needs to explicitly add `-lpthread` if needed.  That's what you fix did in your commit:
https://github.com/edgargabriel/ompi/commit/c583467d81620e8503ef8973a1e5f8d655add219

2. However, your change from `AC_CHECK_FUNCS()` to `OPAL_SEARCH_LIBS_COMPONENT()` removes `HAVE_SEM_INIT` and `HAVE_SEM_OPEN` from `opal_config.h`.  `AC_CHECK_FUNCS()` defines `HAVE_XXX` if it succeeds while `OPAL_SEARCH_LIBS_COMPONENT()` does not.  This causes an error in `sharedfp_sm_file_open.c`: https://github.com/open-mpi/ompi/blob/master/ompi/mca/sharedfp/sm/sharedfp_sm_file_open.c#L179-L193

3. One of the fixes would be calling `AC_CHECK_FUNCS()` in `OPAL_SEARCH_LIBS_COMPONENT()`.  The following is based on your fix.
```patch
diff --git a/ompi/mca/sharedfp/sm/configure.m4 b/ompi/mca/sharedfp/sm/configure.m4
index 7224f15a93..92e1a89692 100644
--- a/ompi/mca/sharedfp/sm/configure.m4
+++ b/ompi/mca/sharedfp/sm/configure.m4
@@ -27,11 +27,19 @@ AC_DEFUN([MCA_ompi_sharedfp_sm_CONFIG],[
 
     sharedfp_sm_happy=no
     AC_CHECK_HEADER([semaphore.h],
-                    [AC_CHECK_FUNCS([sem_open],[sharedfp_sm_happy=yes],[])])
-
+                    [dnl requires potentially pthread library
+                     OPAL_SEARCH_LIBS_COMPONENT([sharedfp_sm],
+                                    [sem_open], [pthread],
+                                    [AC_CHECK_FUNCS([sem_open],
+                                                    [sharedfp_sm_happy="yes"],[])]
+                                                )])
     AC_CHECK_HEADER([semaphore.h],
-        [AC_CHECK_FUNCS([sem_init],[sharedfp_sm_happy=yes],[])])
-
+                    [dnl requires potentially pthread library
+                     OPAL_SEARCH_LIBS_COMPONENT([sharedfp_sm],
+                                    [sem_init], [pthread],
+                                    [AC_CHECK_FUNCS([sem_init],
+                                                    [sharedfp_sm_happy="yes"],[])]
+                                                )])
     AS_IF([test "$sharedfp_sm_happy" = "yes"],
           [$1],
           [$2])
```
Another way is removing `HAVE_SEM_OPEN` and `HAVE_SEM_INIT` from `sharedfp_sm_file_open.c` if they must exist to use this `sharedfp/sm`.

At least this fix worked on my Ubuntu machine.  If this bug is not what you meant, please let me know.

**Comment by edgargabriel** [comment]
@shintaro-iwasaki thank you! I think you identified the problem correctly, I have to admit that I am often lost in the configure script world and am working mostly based on existing examples that I find in the code. Do you want to file a pr against master with this change, and I can test this on my local machine as well?

**Comment by shintaro-iwasaki** [comment]
I am afraid of fixing an issue that is different from what you encountered. If you wouldn't mind, could you copy-paste the following and try it? Once it is confirmed that this fix works in your environment, I would like to create a PR (or you can create a PR if you want).
```
ompi/mca/sharedfp/sm/configure.m4
```

```
AC_DEFUN([MCA_ompi_sharedfp_sm_CONFIG],[
    AC_CONFIG_FILES([ompi/mca/sharedfp/sm/Makefile])

    sharedfp_sm_happy=no
    AC_CHECK_HEADER([semaphore.h],
                    [dnl requires potentially pthread library
                     OPAL_SEARCH_LIBS_COMPONENT([sharedfp_sm],
                                    [sem_open], [pthread],
                                    [AC_CHECK_FUNCS([sem_open],
                                                    [sharedfp_sm_happy="yes"],[])]
                                                )])
    AC_CHECK_HEADER([semaphore.h],
                    [dnl requires potentially pthread library
                     OPAL_SEARCH_LIBS_COMPONENT([sharedfp_sm],
                                    [sem_init], [pthread],
                                    [AC_CHECK_FUNCS([sem_init],
                                                    [sharedfp_sm_happy="yes"],[])]
                                                )])
    AS_IF([test "$sharedfp_sm_happy" = "yes"],
          [$1],
          [$2])
])dnl
```

**Comment by edgargabriel** [comment]
@shintaro-iwasaki I can confirm that with your revamped configure script the sharedfp/sm component configures correctly, compiles and runs correctly again. Thank you!

**Comment by shintaro-iwasaki** [comment]
@edgargabriel Thank you for checking the fix! I created a PR: #8545.
